File: schemas/loyalty-rewards-program/convex/loyalty_members.ts
// loyalty_members: Enrollment of a user in a loyalty program with cached point balances.
// See README.md for full design rationale.
import { defineTable } from "convex/server";
import { v } from "convex/values";

export const loyaltyMembers = defineTable({
  programId: v.id("loyalty_programs"),
  userId: v.id("users"),
  memberNumber: v.string(),
  status: v.union(
    v.literal("active"),
    v.literal("suspended"),
    v.literal("banned")
  ),
  pointsBalance: v.number(),
  pointsPending: v.number(),
  lifetimePoints: v.number(),
  pointsRedeemed: v.number(),
  pointsExpired: v.number(),
  enrolledAt: v.number(),
  suspendedAt: v.optional(v.number()),
  metadata: v.optional(v.any()),
  updatedAt: v.number(),
})
  .index("by_program_id_and_user_id", ["programId", "userId"])
  .index("by_member_number", ["memberNumber"])
  .index("by_user_id", ["userId"])
  .index("by_status", ["status"])
  .index("by_points_balance", ["pointsBalance"]);

File: schemas/loyalty-rewards-program/spacetimedb/loyalty_members.rs
// loyalty_members: Enrollment of a user in a loyalty program with cached point balances.
// See README.md for full design rationale.
use spacetimedb::Timestamp;

#[derive(SpacetimeType, Clone)]
pub enum LoyaltyMemberStatus {
    Active,    // type: String
    Suspended,
    Banned,
}

#[spacetimedb::table(name = loyalty_members, public)]
pub struct LoyaltyMember {
    #[primary_key]
    pub id: String, // UUID

    #[index(btree)]
    pub program_id: String, // UUID, FK → loyalty_programs.id (cascade delete)
    #[index(btree)]
    pub user_id: String, // UUID, FK → users.id (cascade delete)
    // Composite unique: (program_id, user_id) — enforce in reducer logic
    #[unique]
    pub member_number: String,
    #[index(btree)]
    pub status: LoyaltyMemberStatus,
    #[index(btree)]
    pub points_balance: i32, // default 0
    pub points_pending: i32, // default 0
    pub lifetime_points: i32, // default 0
    pub points_redeemed: i32, // default 0
    pub points_expired: i32, // default 0
    pub enrolled_at: Timestamp,
    pub suspended_at: Option<Timestamp>,
    pub metadata: Option<String>, // JSON stored as string
    pub created_at: Timestamp,
    pub updated_at: Timestamp,
}
